src/midi/midi_mt32.cpp

```diff
@@ -483,111 +623,87 @@ static std::set<const LASynthModel*> populate_available_models(
-	                                                &mt32_207_model,
-	                                                &mt32_206_model,
-	                                                &mt32_203_model};
-
-	const size_t max_dir_width = get_max_dir_width(models_without_aliases,
-	                                               indent, column_delim);
-	const size_t truncated_dir_width = max_dir_width - strlen(trailing_dots);
-	assert(truncated_dir_width < max_dir_width);
+	constexpr char indent[]       = "  ";
+
+	const std::vector<const LASynthModel*> mt32_model_list = {&mt32_104_model,
+	                                                          &mt32_105_model,
+	                                                          &mt32_106_model,
+	                                                          &mt32_107_model,
+	                                                          &mt32_bluer_model,
+	                                                          &mt32_203_model,
+	                                                          &mt32_204_model,
+	                                                          &mt32_206_model,
+	                                                          &mt32_207_model};
+
+	const std::vector<const LASynthModel*> cm32_model_list = {
+	        &cm32l_100_model, &cm32l_102_model, &cm32ln_100_model};
+
+	const size_t max_dir_width = get_max_dir_width(indent, column_delim);
 
 	// Get the set of directories and the models they support
-	dirs_with_models_t dirs_with_models;
+	DirsWithModels dirs_with_models;
 	const auto available_models = populate_available_models(GetService(),
 	                                                        dirs_with_models);
+
 	if (available_models.empty()) {
 		caller->WriteOut("%s%s\n", indent, MSG_Get("MT32_NO_SUPPORTED_MODELS"));
 		return MIDI_RC::OK;
 	}
 
-	// Text colors and highlight selector
-	constexpr char gray[]    = "\033[30;1m";
-	constexpr char nocolor[] = "\033[0m";
+	auto highlight_model = [&](const LASynthModel* model,
+	                           const char* display_name) -> std::string {
+		constexpr auto darkgray = "[color=dark-gray]";
+		constexpr auto green    = "[color=light-green]";
+		constexpr auto reset    = "[reset]";
 
-	auto get_highlight = [&](bool is_missing) {
-		return is_missing ? gray : nocolor;
-	};
+		const bool is_missing = (available_models.find(model) ==
+		                         available_models.end());
 
-	// Print the header row of all models
-	const std::string dirs_padding(max_dir_width, ' ');
-	caller->WriteOut("%s%s", indent, dirs_padding.c_str());
-	for (const auto &model : models_without_aliases) {
-		const bool is_missing = (available_models.find(model) == available_models.end());
-		const auto color = get_highlight(is_missing);
-		caller->WriteOut("%s%s%s%s", color, model->GetVersion(),
-		                 nocolor, column_delim);
-	}
-	caller->WriteOut("\n");
+		const auto is_active = (model_and_dir->first == model);
```

Comment:
If `MIXER /LISTMIDI` is called before MT32 is initialized, the std::optional `model_and_dir` will still be empty (this generates an assert on Linux).

Tested this locally:

```c++
const auto is_active = (model_and_dir && model_and_dir->first == model);
````

Reply:
I don't see how that can happen; for me `MIXER /LISTMIDI` prints nothing when `mididevice` is set to `none` in the config, and it works as expected when it's set to `fluidsynth`. That's on macOS. Not sure how you can trigger the assert?

But I can add what you proposed to make things generally more robust, I guess.

Reply:
Done.

Reply:
> Not sure how you can trigger the assert?

On Linux, it at least comes from deep in the STL, but we can just check it before dereferencing (in the original version):

```diff
+ assert(model_and_dir);
+ const auto is_active = (model_and_dir->first == model);
```

#### Before calling `mixer /listmidi`

<img width="1479" alt="before-mixer-listmidi" src="https://github.com/dosbox-staging/dosbox-staging/assets/1557255/8f4ee340-6ee9-489e-872e-514de50d918c">

#### After calling `mixer /listmidi`

<img width="1479" alt="after-mixer-listmidi" src="https://github.com/dosbox-staging/dosbox-staging/assets/1557255/c862beba-76bc-430c-8bd8-07c8354cbd8e">

It seems like clang (even debug builds!) might be missing those internal checks, so it barrels on.. I really don't know how it's giving back a valid but empty `std::pair` value 😆

Reply:
@kcgen Rebased to main again. So are we good now then? 😄